pkg/updater/jobparser.go

```diff
@@ -0,0 +1,337 @@
+/* Copyright (c) 2016 PaddlePaddle Authors All Rights Reserved.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+	 limitations under the License. */
+
+package updater
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1beta1 "k8s.io/api/extensions/v1beta1"
+	apiresource "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	paddlev1 "github.com/paddlepaddle/edl/pkg/apis/paddlepaddle/v1"
+)
+
+const (
+	imagePullPolicy = "Always"
+)
+
+// JobParser is a interface can parse given simple TrainingJob struct to
+// an integrated TrainingJob
+type JobParser interface {
```

Comment:
Do we need this interface?

My personal rule of thumb is not to use interface unless necessary, here are the common reason for using interface:

1. to avoid circular package dependency.
1. multiple implementations of an interface.
1. [dependency inversion principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle): it is about inverting dependency across different packages, so only necessary when multiple packages are involved.

I typically start with one implementation without interface, when there are actually multiple implementations, create an interface. Typically there is never a need for multiple implementations to begin with.

Some related articles: https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8